bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/ModelSyncGradlePlugin.kt

```diff
@@ -0,0 +1,234 @@
+import org.modelix.model.sync.bulk.gradle.tasks.GenerateAntScriptForMps
+import org.modelix.model.sync.bulk.gradle.tasks.ImportIntoModelServer
+import org.modelix.model.sync.bulk.gradle.tasks.ImportIntoMps
+import org.modelix.model.sync.bulk.gradle.tasks.ValidateSyncSettings
+import java.io.File
+import java.util.Properties
+
+class ModelSyncGradlePlugin : Plugin<Project> {
+
+    private lateinit var settings: ModelSyncGradleSettings
+
+    override fun apply(project: Project) {
+        settings = project.extensions.create("modelSync", ModelSyncGradleSettings::class.java)
+        getBaseDir(project).mkdirs()
+
+        project.afterEvaluate {
+            val validateSyncSettings = project.tasks.register("validateSyncSettings", ValidateSyncSettings::class.java) {
+                settings.taskDependencies.forEach { dependency ->
+                    it.dependsOn(dependency)
+                }
+                it.settings.set(settings)
+            }
+            val modelixCoreVersion = readModelixCoreVersion() ?: throw RuntimeException("modelix.core version not found")
+            val antDependencies = project.configurations.create("model-sync-ant-dependencies")
+            project.dependencies.add(antDependencies.name, "org.apache.ant:ant-junit:1.10.12")
+
+            val mpsDependencies = project.configurations.create("modelSyncMpsDependencies")
+            project.dependencies.add(mpsDependencies.name, "org.modelix.mps:bulk-model-sync-solution:$modelixCoreVersion")
+
+            val copyMpsDependencies = project.tasks.register("copyMpsDependencies", Sync::class.java) { sync ->
+                sync.dependsOn(validateSyncSettings)
+                val src = mpsDependencies.resolve().map { project.zipTree(it) }
+                val target = getDependenciesDir(project).apply { mkdirs() }
+                sync.from(src)
+                sync.into(target)
+            }
+
+            settings.syncDirections.forEach {
+                registerTasksForSyncDirection(it, project, copyMpsDependencies)
+            }
+        }
+    }
+
+    private fun registerTasksForSyncDirection(
+        syncDirection: SyncDirection,
+        project: Project,
+        previousTask: TaskProvider<*>,
+    ) {
+        val baseDir = project.buildDir.resolve("model-sync").apply { mkdirs() }
+        val jsonDir = baseDir.resolve(syncDirection.name).apply { mkdir() }
+        val sourceTask = when (syncDirection.source) {
+            is LocalSource -> registerTasksForLocalSource(syncDirection, project, previousTask, jsonDir)
+            is ServerSource -> registerTasksForServerSource(syncDirection, project, previousTask, jsonDir)
+            else -> error("Unknown sync direction source")
+        }
+
+        when (syncDirection.target) {
+            is LocalTarget -> registerTasksForLocalTarget(syncDirection, project, sourceTask, jsonDir)
+            is ServerTarget -> registerTasksForServerTarget(syncDirection, project, sourceTask, jsonDir)
```

Comment:
If the targets were provided to the `register*` method as arguments, the need for casting inside those methods could be eliminated.

Reply:
This would just move the cast to the invocation.
Smart cast is unavailable because the field is mutable.

Reply:
You can assign it to a `val` in the function to avoid that problem.